File: python/logger.py

from sys import stderr

class Logger:
    """A logger class to control output to STDERR."""
    instance = None
    def __init__(self, debug_level = 0):
        self.debug_level = debug_level
    def echo(self, *args, level = 1, flush = False, **key):
        """
        Prints out the arguments, as though passed to `print', if and only if
        the debug level is greater than or equal to the level passed in. If the
        flush argument is true, the output is flushed immediately.
        """
        if self.debug_level >= level:
            print(*args, file = stderr, **key)
            if flush:
                stderr.flush()
    @staticmethod
    def get():
        if Logger.instance is None:
            Logger.instance = Logger()
        return Logger.instance

def set_global_debug_level(lvl):
    """Sets the debug level of the default logger provided in Logger."""
    Logger.get().debug_level = lvl

def echo(*args, level = 1, flush = False, **key):
    """Echoes to the default logger."""
    Logger.get().echo(*args, level = level, flush = flush, **key)

File: ruby/logger.rb

# A Logger class which conditionally outputs debug information to STDERR, depending on the debug level set
# when the program was invoked.
class Logger

  # Returns the default logger instance. Note that Logger is not a singleton class; new instances can
  # be freely created; this is merely a default instance provided for programmer convenience.
  def self.instance
    @@instance ||= Logger.new
  end

  # Echo the string to the default logger instance, using #echo.
  def self.echo(level, str)
    Logger.instance.echo level, str
  end

  def initialize(debug_level = 0)
    @level = debug_level
  end

  def debug_level
    @level
  end

  def debug_level=(val)
    @level = val.to_i
  end

  # Outputs the information given in the string argument if (and only if) the current debug level for this
  # logger is greater than or equal to the supplied level.
  def echo(level, str)
    STDERR.puts "#{$$} [3] #{str}" if debug_level >= level
  end

end
